implementation/BeforeInstallPromptEvent.js

```diff
@@ -0,0 +1,143 @@
+/*globals BeforeInstallPromptEvent, DOMException*/
+"use strict";
+const internalSlots = new WeakMap();
+const installProcesses = [];
+const AppBannerPromptOutcome = new Set([
+  "accepted",
+  "dismissed",
+]);
+
+class BeforeInstallPromptEvent extends Event {
+  constructor(typeArg, eventInit) {
+    // WebIDL Guard. Wont be in spec in spec as it's all handled by WebIDL.
+    if (arguments.length === 0) {
+      throw new TypeError("Not enough arguments. Expected at least 1.");
+    }
+    const initType = typeof eventInit;
+    if (arguments.length === 2 && initType !== "undefined" && initType !== "object") {
+      throw new TypeError("Value can't be converted to a dictionary.");
+    }
+    super(typeArg, Object.assign({ cancelable: true }, eventInit));
+
+    if (eventInit && typeof eventInit.userChoice !== "undefined" && !AppBannerPromptOutcome.has(String(eventInit.userChoice))) {
+      const msg = `The provided value '${eventInit.userChoice}' is not a valid` +
+        "enum value of type AppBannerPromptOutcome.";
+      throw new TypeError(msg);
+    }
+    // End WebIDL guard.
+    const internal = {
+      didPrompt: false,
+    };
+
+    internal.userChoice = new Promise((resolve) => {
+      internal.userChoiceHandlers = {
+        resolve,
+      };
+      if (eventInit && "userChoice" in eventInit) {
+        resolve(eventInit.userChoice);
+      }
+    });
+    internalSlots.set(this, internal);
+  }
+  prompt() {
+    if (internalSlots.get(this).didPrompt) {
+      const msg = ".prompt() can only be called once.";
+      throw new DOMException(msg, "InvalidStateError");
+    } else {
+      internalSlots.get(this).didPrompt = true;
+    }
+
+    if (this.isTrusted === false) {
+      const msg = "Untrusted events can't call prompt().";
+      throw new DOMException(msg, "NotAllowedError");
+    }
+
+    if (this.defaultPrevented === false) {
+      const msg = ".prompt() needs to be called after .preventDefault()";
+      throw new DOMException(msg, "InvalidStateError");
+    }
+
+    (async function task() {
+      const promptOutcome = await showInstallPrompt();
+      internalSlots.get(this).userChoiceHandlers.resolve(promptOutcome);
+    }.bind(this)())
+  }
+
+  get userChoice() {
+    return internalSlots.get(this).userChoice;
+  }
+}
+
+async function notifyBeforeInstallPrompt(element) {
+  await trackReadyState();
+  if (installProcesses.length) {
+    return;
+  }
+  const event = new BeforeInstallPromptEvent("beforeinstallprompt");
+  window.dispatchEvent(event);
+  if (!event.defaultPrevented) {
+    await showInstallPrompt(element);
+  }
+}
+
+function trackReadyState() {
+  return new Promise((resolve) => {
+    if (document.readyState === "complete") {
+      return resolve();
+    }
+    document.addEventListener("readystatechange", () => {
+      if (document.readyState === "complete") {
+        resolve();
+      }
+    });
+  });
+}
+
+window.addEventListener("beforeinstallprompt", async function(event) {
```

Comment:
It's weird that this code is here. I think of this file as being a polyfill for `BeforeInstallPromptEvent`, so it provides the event object as well as the methods for simulating the system sending those events (`notifyBeforeInstallPrompt`). It shouldn't also have a handler for the event which automatically calls preventDefault and prompt.

It's even weirder because `index.html` has its own BIP handler which may race with this one.